fling/component/client/fedmoon_client.py

```diff
@@ -0,0 +1,95 @@
+import copy
+
+import torch
+import torch.nn as nn
+
+from fling.utils.registry_utils import CLIENT_REGISTRY
+from .base_client import BaseClient
+
+
+@CLIENT_REGISTRY.register('fedmoon_client')
+class FedMOONClient(BaseClient):
+    r"""
+    Overview:
+        This class is the base implementation of client of FedMOON introduced in: Model-Contrastive Federated Learning
+     <link https://openaccess.thecvf.com/content/CVPR2021/papers/Li_Model-Contrastive_Federated_Learning_CVPR_2021_paper.pdf link>.
+    """
+
+    def __init__(self, *args, **kwargs):
+        super(FedMOONClient, self).__init__(*args, **kwargs)
+        # The weight of fedmoon loss.
+        self.mu = self.args.learn.mu
+        # The temperature parameter of fedmoon.
+        self.t = self.args.learn.temperature
+        # The variable to store the global model.
+        self.glob_model = None
+        # The variable to store the previous models.
+        self.prev_models = []
+
+    def _store_prev_model(self, model: nn.Module) -> None:
+        r"""
+        Overview:
+            Store the prev model for fedmoon loss calculation.
+        """
+        self.prev_models.append(copy.deepcopy(model.to(self.device)))
+
+    def _store_global_model(self, model: nn.Module) -> None:
+        r"""
+        Overview:
+            Store the global model for fedmoon loss calculation.
+        """
+        self.glob_model = copy.deepcopy(model.to(self.device))
+
+    def train_step(self, batch_data, criterion, monitor, optimizer):
+        r"""
+        Overview:
+            Training step. The loss of fedmoon should be added to the original loss.
+        """
+        batch_x, batch_y = batch_data['x'], batch_data['y']
+        z, o = self.model(batch_x)
+        main_loss = criterion(o, batch_y)
+        # Calculate fedmoon loss.
+        cos = nn.CosineSimilarity(dim=-1)
+        z_glob, _ = self.glob_model(batch_x)
+        z_i = cos(z, z_glob)
+        logits = z_i.reshape(-1, 1)
+        for prev_model in self.prev_models:
```

Comment:
I don't think it is a good idea to forward calculate all ``self.prev_models``. As the training round is getting larger, ``self.prev_models`` will finally have infinite items. I think a queue with fixed length is better.